tools/stonith_admin.c

```diff
@@ -572,17 +572,49 @@ main(int argc, char **argv)
                 devices = stonith_key_value_add(devices, NULL, optarg);
                 break;
             case 'o':
-                crm_info("Scanning: -o %s", optarg);
-                rc = sscanf(optarg, "%m[^=]=%m[^=]", &name, &value);
-                if (rc != 2) {
-                    crm_err("Invalid option: -o %s", optarg);
-                    ++argerr;
-                } else {
-                    crm_info("Got: '%s'='%s'", name, value);
-                    params = stonith_key_value_add(params, name, value);
+                {
+#ifdef SSCANF_HAS_M
+                    crm_info("Scanning: -o %s", optarg);
+                    rc = sscanf(optarg, "%m[^=]=%m[^=]", &name, &value);
+#else
+                    char *sep = NULL;
+
+                    crm_info("Scanning: -o %s", optarg);
+
+                    sep = strstr(optarg, "=");
+                    if (sep != NULL) {
+                        name = strndup(optarg, sep-optarg);
+
+                        /* If the last char in optarg is =, the user gave no
+                         * value for the option.  Leave it as NULL.
+                         */
+                        if (*(sep+1) != '\0') {
+                            /* Skip over the '=' that the first strstr left
+                             * us on.  See if there's a second '='.
+                             */
+                            char *sep2 = strstr(sep+1, "=");
+
+                            if (sep2 != NULL) {
+                                /* There was a second '='.  Copy everything up
+                                 * to it into the value.  Throw away the rest.
+                                 */
+                                value = strndup(sep+1, sep2-sep-1);
+                            } else {
+                                value = strdup(sep+1);
+                            }
+                        }
+                    }
+#endif
```

Comment:
Let's put this (minus the crm_info()) into lib/common/strings.c (declared in include/crm/common/util.h) as something like pcmk_scan_nvpair(). Probably worthwhile to set name/value to NULL at the beginning of the function for when it fails.

We can return the number of items matched on success (1 or 2) and -errno otherwise. (I don't know if you've seen our -errno convention yet -- not literally -errno but the negative of some error code whether a system errno or one of the pcmk_err_* codes, which we can add to as needed.)

What's the point of checking for a second '='?

Reply:
I'll move it on over.

There's not really any point to checking for a second '='.  I was just trying to preserve existing behavior.  sscanf is using '%m[^=] for the value, so anything after a second '=' is going to get dropped.  If you'd like, I can ditch all that code (it was annoying to write) and fix the sscanf case to just pull in the whole rest of optarg into value.

Reply:
Yes, that sounds good. I'm guessing the second [^=] was an unintentional copy-paste.